Kudu.Services/Arm/ArmUtils.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using Kudu.Contracts.Infrastructure;
+
+namespace Kudu.Services.Arm
+{
+    public static class ArmUtils
+    {
+        public const string GeoLocationHeaderKey = "x-ms-geo-location";
+
+        public static object AddEnvelopeOnArmRequest<T>(T namedObject, HttpRequestMessage request) where T : INamedObject
+        {
+            if (IsArmRequest(request))
+            {
+                return Create(namedObject, request);
+            }
+
+            return namedObject;
+        }
+
+        public static object AddEnvelopeOnArmRequest<T>(IEnumerable<T> namedObjects, HttpRequestMessage request) where T : INamedObject
+        {
+            if (IsArmRequest(request))
+            {
+                return Create(namedObjects, request);
+            }
+
+            return namedObjects;
+        }
+
+        public static bool IsArmRequest(HttpRequestMessage request)
+        {
+            return request != null &&
+                   request.Headers != null &&
+                   request.Headers.Contains(GeoLocationHeaderKey);
+        }
+
+        private static ArmListEntry<T> Create<T>(IEnumerable<T> objects, HttpRequestMessage request) where T : INamedObject
+        {
+            return new ArmListEntry<T>
+            {
+                Value = objects.Select(entry => Create(entry, request, isChild: true))
+            };
+        }
+
+        private static ArmEntry<T> Create<T>(T o, HttpRequestMessage request, bool isChild = false) where T : INamedObject
+        {
+            var armEntry = new ArmEntry<T>()
+            {
+                Properties = o
+            };
+
+            // In Azure ARM requests, the referrer is the current id
+            Uri referrer = request.Headers.Referrer;
+            if (referrer != null)
```

Comment:
This is one of the contract that it came from CSM.   Maybe we should add this check in IsArmRequest - and no need to check here.